// tstl/container/TreeMultiMap.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
//================================================================ 
/** @module std */
//================================================================
var MultiMap_1 = require("../base/container/MultiMap");
var _ITreeContainer_1 = require("../base/container/_ITreeContainer");
var _MultiMapTree_1 = require("../base/tree/_MultiMapTree");
var MapIterator_1 = require("../base/iterator/MapIterator");
var Entry_1 = require("../utility/Entry");
/**
 * Multiple-key Map based on Tree.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var TreeMultiMap = /** @class */ (function (_super) {
    __extends(TreeMultiMap, _super);
    function TreeMultiMap() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = _super.call(this) || this;
        _ITreeContainer_1._Construct.apply(void 0, __spread([_this, TreeMultiMap,
            function (comp) {
                _this.tree_ = new _MultiMapTree_1._MultiMapTree(_this, comp);
            }], args));
        return _this;
    }
    /* ---------------------------------------------------------
        ASSIGN & CLEAR
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.clear = function () {
        _super.prototype.clear.call(this);
        this.tree_.clear();
    };
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.swap = function (obj) {
        var _a;
        // SWAP CONTENTS
        _super.prototype.swap.call(this, obj);
        // SWAP RB-TREE
        _MultiMapTree_1._MultiMapTree._Swap_source(this.tree_, obj.tree_);
        _a = __read([obj.tree_, this.tree_], 2), this.tree_ = _a[0], obj.tree_ = _a[1];
    };
    /* =========================================================
        ACCESSORS
    ========================================================= */
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.find = function (key) {
        var node = this.tree_.nearest_by_key(key);
        if (node === null || this.tree_.key_eq()(node.value.first, key) === false)
            return this.end();
        else
            return node.value;
    };
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.count = function (key) {
        var it = this.find(key);
        var ret = 0;
        for (; !it.equals(this.end()) && this.tree_.key_eq()(it.first, key); it = it.next())
            ++ret;
        return ret;
    };
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.key_comp = function () {
        return this.tree_.key_comp();
    };
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.value_comp = function () {
        return this.tree_.value_comp();
    };
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.lower_bound = function (key) {
        return this.tree_.lower_bound(key);
    };
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.upper_bound = function (key) {
        return this.tree_.upper_bound(key);
    };
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.equal_range = function (key) {
        return this.tree_.equal_range(key);
    };
    /**
     * @hidden
     */
    TreeMultiMap.prototype._Key_eq = function (x, y) {
        return this.tree_.key_eq()(x, y);
    };
    /* =========================================================
        ELEMENTS I/O
            - INSERT
            - POST-PROCESS
    ============================================================
        INSERT
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.emplace = function (key, val) {
        // FIND POSITION TO INSERT
        var it = this.upper_bound(key);
        // ITERATOR TO RETURN
        it = this.data_.insert(it, new Entry_1.Entry(key, val));
        this._Handle_insert(it, it.next()); // POST-PROCESS
        return it;
    };
    /**
     * @inheritDoc
     */
    TreeMultiMap.prototype.emplace_hint = function (hint, key, val) {
        var _this = this;
        var elem = new Entry_1.Entry(key, val);
        return _ITreeContainer_1._Emplace_hint(this, hint, elem, this.data_, this._Handle_insert.bind(this), function () { return _this.emplace(key, val); });
    };
    /**
     * @hidden
     */
    TreeMultiMap.prototype._Insert_by_range = function (first, last) {
        for (var it = first; !it.equals(last); it = it.next())
            this.emplace(it.value.first, it.value.second);
    };
    /* ---------------------------------------------------------
        POST-PROCESS
    --------------------------------------------------------- */
    /**
     * @hidden
     */
    TreeMultiMap.prototype._Handle_insert = function (first, last) {
        for (; !first.equals(last); first = first.next())
            this.tree_.insert(first);
    };
    /**
     * @hidden
     */
    TreeMultiMap.prototype._Handle_erase = function (first, last) {
        for (; !first.equals(last); first = first.next())
            this.tree_.erase(first);
    };
    return TreeMultiMap;
}(MultiMap_1.MultiMap));
exports.TreeMultiMap = TreeMultiMap;
(function (TreeMultiMap) {
    // BODY
    TreeMultiMap.Iterator = MapIterator_1.MapIterator;
    TreeMultiMap.ReverseIterator = MapIterator_1.MapReverseIterator;
    // BODY
    TreeMultiMap.iterator = TreeMultiMap.Iterator;
    TreeMultiMap.reverse_iterator = TreeMultiMap.ReverseIterator;
})(TreeMultiMap = exports.TreeMultiMap || (exports.TreeMultiMap = {}));
exports.TreeMultiMap = TreeMultiMap;
exports.multimap = TreeMultiMap;
