// tstl/container/TreeMultiSet.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
//================================================================ 
/** @module std */
//================================================================
var MultiSet_1 = require("../base/container/MultiSet");
var _ITreeContainer_1 = require("../base/container/_ITreeContainer");
var _MultiSetTree_1 = require("../base/tree/_MultiSetTree");
var SetIterator_1 = require("../base/iterator/SetIterator");
/**
 * Multiple-key Set based on Tree.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var TreeMultiSet = /** @class */ (function (_super) {
    __extends(TreeMultiSet, _super);
    function TreeMultiSet() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var _this = _super.call(this) || this;
        _ITreeContainer_1._Construct.apply(void 0, __spread([_this, TreeMultiSet,
            function (comp) {
                _this.tree_ = new _MultiSetTree_1._MultiSetTree(_this, comp);
            }], args));
        return _this;
    }
    /* ---------------------------------------------------------
        ASSIGN & CLEAR
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.clear = function () {
        _super.prototype.clear.call(this);
        this.tree_.clear();
    };
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.swap = function (obj) {
        var _a;
        // SWAP CONTENTS
        _super.prototype.swap.call(this, obj);
        // SWAP RB-TREE
        _MultiSetTree_1._MultiSetTree._Swap_source(this.tree_, obj.tree_);
        _a = __read([obj.tree_, this.tree_], 2), this.tree_ = _a[0], obj.tree_ = _a[1];
    };
    /* =========================================================
        ACCESSORS
    ========================================================= */
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.find = function (key) {
        var node = this.tree_.nearest_by_key(key);
        if (node === null || this.tree_.key_eq()(node.value.value, key) === false)
            return this.end();
        else
            return node.value;
    };
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.count = function (key) {
        var it = this.find(key);
        var ret = 0;
        for (; !it.equals(this.end()) && this.tree_.key_eq()(it.value, key); it = it.next())
            ++ret;
        return ret;
    };
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.key_comp = function () {
        return this.tree_.key_comp();
    };
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.value_comp = function () {
        return this.tree_.key_comp();
    };
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.lower_bound = function (key) {
        return this.tree_.lower_bound(key);
    };
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.upper_bound = function (key) {
        return this.tree_.upper_bound(key);
    };
    /**
     * @inheritDoc
     */
    TreeMultiSet.prototype.equal_range = function (key) {
        return this.tree_.equal_range(key);
    };
    /**
     * @hidden
     */
    TreeMultiSet.prototype._Key_eq = function (x, y) {
        return this.tree_.key_eq()(x, y);
    };
    /* =========================================================
        ELEMENTS I/O
            - INSERT
            - POST-PROCESS
    ============================================================
        INSERT
    --------------------------------------------------------- */
    /**
     * @hidden
     */
    TreeMultiSet.prototype._Insert_by_key = function (key) {
        // FIND POSITION TO INSERT
        var it = this.upper_bound(key);
        // ITERATOR TO RETURN
        it = this.data_.insert(it, key);
        this._Handle_insert(it, it.next()); // POST-PROCESS
        return it;
    };
    /**
     * @hidden
     */
    TreeMultiSet.prototype._Insert_by_hint = function (hint, key) {
        var _this = this;
        return _ITreeContainer_1._Emplace_hint(this, hint, key, this.data_, this._Handle_insert.bind(this), function () { return _this._Insert_by_key(key); });
    };
    /**
     * @hidden
     */
    TreeMultiSet.prototype._Insert_by_range = function (first, last) {
        for (; !first.equals(last); first = first.next())
            this._Insert_by_key(first.value);
    };
    /* ---------------------------------------------------------
        POST-PROCESS
    --------------------------------------------------------- */
    /**
     * @hidden
     */
    TreeMultiSet.prototype._Handle_insert = function (first, last) {
        for (; !first.equals(last); first = first.next())
            this.tree_.insert(first);
    };
    /**
     * @hidden
     */
    TreeMultiSet.prototype._Handle_erase = function (first, last) {
        for (; !first.equals(last); first = first.next())
            this.tree_.erase(first);
    };
    return TreeMultiSet;
}(MultiSet_1.MultiSet));
exports.TreeMultiSet = TreeMultiSet;
(function (TreeMultiSet) {
    // BODY
    TreeMultiSet.Iterator = SetIterator_1.SetIterator;
    TreeMultiSet.ReverseIterator = SetIterator_1.SetReverseIterator;
    // BODY
    TreeMultiSet.iterator = TreeMultiSet.Iterator;
    TreeMultiSet.reverse_iterator = TreeMultiSet.ReverseIterator;
})(TreeMultiSet = exports.TreeMultiSet || (exports.TreeMultiSet = {}));
exports.TreeMultiSet = TreeMultiSet;
exports.multiset = TreeMultiSet;
