// tstl/experimental/thread/TimedSemaphore.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var HashMap_1 = require("../../container/HashMap");
var LogicError_1 = require("../../exception/LogicError");
var RuntimeError_1 = require("../../exception/RuntimeError");
var global_1 = require("../../thread/global");
/**
 * Timed semaphore.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var TimedSemaphore = /** @class */ (function () {
    /* ---------------------------------------------------------
        CONSTRUCTORS
    --------------------------------------------------------- */
    /**
     * Construct from section size.
     *
     * @param size Number of maximum sections lockable.
     */
    function TimedSemaphore(size) {
        this.locked_count_ = 0;
        this.hold_count_ = 0;
        this.size_ = size;
        this.resolvers_ = new HashMap_1.HashMap();
    }
    /**
     * @inheritDoc
     */
    TimedSemaphore.prototype.size = function () {
        return this.size_;
    };
    /**
     * @hidden
     */
    TimedSemaphore.prototype._Compute_excess_count = function (count) {
        return Math.max(0, Math.min(this.locked_count_, this.size_) + count - this.size_);
    };
    /**
     * @hidden
     */
    TimedSemaphore.prototype._Compute_resolve_count = function (count) {
        return Math.min(count, this.hold_count_);
    };
    /* ---------------------------------------------------------
        ACQURE & RELEASE
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    TimedSemaphore.prototype.lock = function (count) {
        var _this = this;
        if (count === void 0) { count = 1; }
        return new Promise(function (resolve, reject) {
            // VALIDATE PARAMETER
            if (count < 1 || count > _this.size_) {
                reject(new LogicError_1.OutOfRange("Lock count to semaphore is out of its range."));
                return;
            }
            // INCREASE COUNT PROPERTIES
            var exceeded_count = _this._Compute_excess_count(count);
            _this.hold_count_ += exceeded_count;
            _this.locked_count_ += count;
            // BRANCH; KEEP OR GO?
            if (exceeded_count > 0)
                _this.resolvers_.emplace(resolve, {
                    count: exceeded_count,
                    type: 0 /* HOLD */
                });
            else
                resolve();
        });
    };
    /**
     * @inheritDoc
     */
    TimedSemaphore.prototype.try_lock = function (count) {
        if (count === void 0) { count = 1; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                // VALIDATE PARAMETER
                if (count < 1 || count > this.size_)
                    throw new LogicError_1.OutOfRange("Lock count to semaphore is out of its range.");
                // ALL OR NOTHING
                if (this.locked_count_ + count > this.size_)
                    return [2 /*return*/, false];
                this.locked_count_ += count;
                return [2 /*return*/, true];
            });
        });
    };
    /**
     * @inheritDoc
     */
    TimedSemaphore.prototype.unlock = function (count) {
        if (count === void 0) { count = 1; }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // VALIDATE PARAMETER
                        if (count < 1 || count > this.size_)
                            throw new LogicError_1.OutOfRange("Unlock count to semaphore is out of its range.");
                        else if (count > this.locked_count_)
                            throw new RuntimeError_1.RangeError("Number of unlocks to semaphore is greater than its locks.");
                        // DO RELEASE
                        this.locked_count_ -= count;
                        return [4 /*yield*/, this._Unlock(count)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @hidden
     */
    TimedSemaphore.prototype._Unlock = function (resolved_count) {
        return __awaiter(this, void 0, void 0, function () {
            var it /*Iterator*/, props;
            return __generator(this, function (_a) {
                // COMPUTE PROPERTY
                resolved_count = this._Compute_resolve_count(resolved_count);
                this.hold_count_ -= resolved_count;
                while (resolved_count !== 0) {
                    it = this.resolvers_.begin();
                    props = it.second;
                    if (props.count > resolved_count) {
                        props.count -= resolved_count;
                        resolved_count = 0;
                    }
                    else {
                        // POP AND DECREAE COUNT FIRST
                        resolved_count -= props.count;
                        this.resolvers_.erase(it);
                        // INFORM UNLOCK
                        if (props.type === 0 /* HOLD */)
                            it.first();
                        else
                            it.first(true);
                    }
                }
                return [2 /*return*/];
            });
        });
    };
    /* ---------------------------------------------------------
        TIMED ACQUIRE
    --------------------------------------------------------- */
    /**
     * Try lock sections until timeout.
     *
     * @param ms The maximum miliseconds for waiting.
     * @param count Count to lock.
     * @return Whether succeded to lock or not.
     */
    TimedSemaphore.prototype.try_lock_for = function (ms, count) {
        if (count === void 0) { count = 1; }
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        // VALIDATE PARAMETER
                        if (count < 1 || count > _this.size_) {
                            reject(new LogicError_1.OutOfRange("Lock count to semaphore is out of its range."));
                            return;
                        }
                        // INCRESE COUNT PROPERTIES
                        var exceeded_count = _this._Compute_excess_count(count);
                        _this.hold_count_ += exceeded_count;
                        _this.locked_count_ += count;
                        // BRANCH; KEEP OR GO?
                        if (exceeded_count > 0) {
                            // RESERVATE LOCK
                            _this.resolvers_.emplace(resolve, {
                                count: exceeded_count,
                                type: 1 /* KNOCK */
                            });
                            // DO SLEEP
                            global_1.sleep_for(ms).then(function () {
                                var it /*Iterator*/ = _this.resolvers_.find(resolve);
                                if (it.equals(_this.resolvers_.end()) === true)
                                    return; // ALREADY BE RETURNED
                                //----
                                // ADJUSTMENTS
                                //----
                                // ALL OR NOTHING
                                _this.locked_count_ -= count - (exceeded_count - it.second.count);
                                // ERASE RESOLVER
                                _this.hold_count_ -= it.second.count;
                                _this.resolvers_.erase(it);
                                // RELEASE FOR THE NEXT HOLDERS
                                _this._Unlock(it.second.count).then(function () {
                                    // RETURNS
                                    resolve(false);
                                });
                            });
                        }
                        else // SUCCEEDED AT ONCE
                            resolve(true);
                    })];
            });
        });
    };
    /**
     * Try lock sections until time expiration.
     *
     * @param at The maximum time point to wait.
     * @param count Count to lock.
     * @return Whether succeded to lock or not.
     */
    TimedSemaphore.prototype.try_lock_until = function (at, count) {
        if (count === void 0) { count = 1; }
        // COMPUTE MILLISECONDS TO WAIT
        var now = new Date();
        var ms = at.getTime() - now.getTime();
        return this.try_lock_for(ms, count);
    };
    return TimedSemaphore;
}());
exports.TimedSemaphore = TimedSemaphore;
exports.timed_semaphore = TimedSemaphore;
