// tstl/base/iterator/ArrayIterator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ReverseIterator_1 = require("./ReverseIterator");
var comparators_1 = require("../../functional/comparators");
/**
 * Iterator of Array Containers.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var ArrayIterator = /** @class */ (function () {
    /* ---------------------------------------------------------
        CONSTRUCTORS
    --------------------------------------------------------- */
    /**
     * Initializer Constructor.
     *
     * @param source Source container.
     * @param index Index number.
     */
    function ArrayIterator(source, index) {
        this.source_ = source;
        this.index_ = index;
    }
    /**
     * @inheritDoc
     */
    ArrayIterator.prototype.reverse = function () {
        return new ArrayReverseIterator(this);
    };
    /* ---------------------------------------------------------
        ACCESSORS
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    ArrayIterator.prototype.source = function () {
        return this.source_;
    };
    /**
     * @inheritDoc
     */
    ArrayIterator.prototype.index = function () {
        return this.index_;
    };
    Object.defineProperty(ArrayIterator.prototype, "value", {
        /**
         * @inheritDoc
         */
        get: function () {
            return this.source().at(this.index_);
        },
        /**
         * @inheritDoc
         */
        set: function (val) {
            this.source().set(this.index_, val);
        },
        enumerable: true,
        configurable: true
    });
    /* ---------------------------------------------------------
        MOVERS
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    ArrayIterator.prototype.prev = function () {
        return new ArrayIterator(this.source(), this.index_ - 1);
    };
    /**
     * @inheritDoc
     */
    ArrayIterator.prototype.next = function () {
        return new ArrayIterator(this.source(), this.index_ + 1);
    };
    /**
     * @inheritDoc
     */
    ArrayIterator.prototype.advance = function (n) {
        return new ArrayIterator(this.source(), this.index_ + n);
    };
    /* ---------------------------------------------------------
        COMPARES
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    ArrayIterator.prototype.equals = function (obj) {
        return comparators_1.equal_to(this.source_, obj.source_) && this.index_ === obj.index_;
    };
    return ArrayIterator;
}());
exports.ArrayIterator = ArrayIterator;
/**
 * Reverse iterator of Array Containers.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var ArrayReverseIterator = /** @class */ (function (_super) {
    __extends(ArrayReverseIterator, _super);
    /* ---------------------------------------------------------
        CONSTRUCTORS
    --------------------------------------------------------- */
    /**
     * Initializer Constructor.
     *
     * @param base The base iterator.
     */
    function ArrayReverseIterator(base) {
        return _super.call(this, base) || this;
    }
    /**
     * @hidden
     */
    ArrayReverseIterator.prototype._Create_neighbor = function (base) {
        return new ArrayReverseIterator(base);
    };
    /**
     * @inheritDoc
     */
    ArrayReverseIterator.prototype.advance = function (n) {
        return this._Create_neighbor(this.base().advance(-n));
    };
    /* ---------------------------------------------------------
        ACCESSORS
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    ArrayReverseIterator.prototype.index = function () {
        return this.base_.index();
    };
    Object.defineProperty(ArrayReverseIterator.prototype, "value", {
        /**
         * @inheritDoc
         */
        get: function () {
            return this.base_.value;
        },
        /**
         * @inheritDoc
         */
        set: function (val) {
            this.base_.value = val;
        },
        enumerable: true,
        configurable: true
    });
    return ArrayReverseIterator;
}(ReverseIterator_1.ReverseIterator));
exports.ArrayReverseIterator = ArrayReverseIterator;
